src/order/category/Semilattice.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.category.PartialOrder
+import order.hom.lattice
+
+/-!
+# The categories of semilattices
+
+This defines `SemilatticeSup` and `SemilatticeInf`, the categories of sup-semilattices with a bottom
+element and inf-semilattices with a top element.
+
+## References
+
+* [nLab, *semilattice*](https://ncatlab.org/nlab/show/semilattice)
+-/
+
+universes u
+open category_theory
+
+/-- The category of sup-semilattices with a bottom element. -/
+structure SemilatticeSup : Type.{u + 1} :=
+(X : Type.{u})
+[is_semilattice_sup : semilattice_sup X]
+[is_order_bot : order_bot X]
+
+/-- The category of inf-semilattices with a top element. -/
+structure SemilatticeInf : Type.{u + 1} :=
+(X : Type.{u})
+[is_semilattice_inf : semilattice_inf X]
+[is_order_top : order_top X]
+
+namespace SemilatticeSup
+
+instance : has_coe_to_sort SemilatticeSup Type* := ⟨X⟩
+
+attribute [protected] X
```

Comment:
Would you mind putting this directly below the definition of `SemilatticeSup` so we can write `attribute [protected] SemilatticeSup.X` for the sake of readability?